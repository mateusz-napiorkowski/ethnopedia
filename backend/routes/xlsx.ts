import express from "express"

const router = express.Router()

const {
    getXlsxWithAllData,
    getAllCaterories
} = require("../controllers/xlsx")

router.route("/:collectionName").get(getXlsxWithAllData)
router.route("/keys/:collectionName").get(getAllCaterories)

module.exports = router