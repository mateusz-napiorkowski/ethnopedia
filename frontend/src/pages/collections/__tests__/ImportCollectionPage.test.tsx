import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { UserContext } from '../../../providers/UserProvider';
import ImportCollectionPage from '../ImportCollectionPage';
import * as XLSX from 'xlsx';

const queryClient = new QueryClient();
const user = userEvent.setup()

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockUseNavigate
}));

const mockImportDataAsCollection = jest.fn()
jest.mock('../../../api/dataImport', () => ({
   importDataAsCollection: (
        importData: any,
        collectionName: string | undefined,
        description: string,
        jwtToken: string,
        archiveFile: File | undefined,
    ) => mockImportDataAsCollection(importData, collectionName, description, jwtToken, archiveFile)
}));

const exampleCollectionData = {
    _id: '662e928b11674920c8cc0aaa',
    name: 'example collection',
    description: 'example collection description'
}
const jwtToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3Rvd3kiLCJmaXJzdE5hbWUiOiJ0ZXN0b3d5IiwidXN"
    + "lcklkIjoiNjZiNjUwNmZiYjY0ZGYxNjVlOGE5Y2U2IiwiaWF0IjoxNzI0MTg0MTE0LCJleHAiOjE3MjUxODQxMTR9.fzHPaXFMzQTVUf9IdZ0G6oeiaecc"
    + "N-rDSjRS3kApqlA"

const fileData = [
    ['Title', 'Title.Subtitle', 'Title.Subtitle.Subsubtitle', 'Artists'],
    ['title 1', 'subtitle 1',
    'subsubtitle 1', "artist 1"],
]

const newFileData = [
    ['Title', 'Title.Subtitle', 'Title.Subtitle.Subsubtitle', 'Artists'],
    ['new file title 1', 'new file subtitle 1', 'new file subsubtitle 1', "new file artist 1"],
]

const fileDataWithIdsAndFilenames = [
    ['_id', 'nazwy plik√≥w', 'Title', 'Title.Subtitle', 'Title.Subtitle.Subsubtitle', 'Artists'],
    ['1234aaaa46e5db48231024ef', '0:filename.mid', 'title 1', 'subtitle 1', 'subsubtitle 1', "artist 1"],
]

const createXlsxFile = (xlsxData: Array<Array<string>>, fileName: string) => {
    const worksheet = XLSX.utils.aoa_to_sheet(xlsxData);
    const workbook = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(workbook, worksheet, "Sheet1");
    const wbout = XLSX.write(workbook, { bookType: "xlsx", type: "array" });
    const blob = new Blob([wbout], {
        type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    });
    const file = new File([blob], fileName, { type: blob.type });
    return file
}

const renderComponent = (
        path = "/import-collection",
        userContextProps: any = {
            isUserLoggedIn: true,
            firstName: "123",
            userId: "66b6506fbb64df165e8a9ce6",
            jwtToken: jwtToken,
            setUserData: jest.fn()
        }
) => {
    return render(
        <UserContext.Provider value={ userContextProps }>
            <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[path]}>
                <Routes>
                    <Route path={path} element={<ImportCollectionPage/>}/>
                </Routes>  
            </MemoryRouter>
        </QueryClientProvider>
        </UserContext.Provider>  
    );
};

describe("ImportCollectionPage tests", () => {
    beforeEach(() => {
        queryClient.clear();
        jest.clearAllMocks()
    });

    it("should render initial state", () => {           
        const {container, getByLabelText} = renderComponent()

        expect(container).toMatchSnapshot()
        expect(getByLabelText("import-data")).toBeDisabled()
    })

    it("should navigate to previous page after cancel button is clicked", async () => {           
        const {getByText} = renderComponent()
        const cancelButton = getByText(/anuluj/i)

        await user.click(cancelButton)

        expect(mockUseNavigate).toHaveBeenCalledWith(-1)
    })

    it("should update upload component and show loaded categories configuration menu with correct initial category structure after file is loaded", async () => {           
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload")
        const file = createXlsxFile(fileData, "example.xlsx")      

        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );

        expect(container).toMatchSnapshot()
    })

    
    it("should unload file data after remove file to upload button is clicked and hide categories configuration menu", async () => {           
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload")
        const file = createXlsxFile(fileData, "example.xlsx")
            
        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );
        await user.click(getByLabelText("remove-file-to-load"))
        
        expect(container).toMatchSnapshot()
    })

    it("should reload file data after file is loaded then removed and loaded again", async () => {
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload")
        const file = createXlsxFile(fileData, "example.xlsx")

        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );
        await user.click(getByLabelText("remove-file-to-load"))
        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );

        expect(container).toMatchSnapshot()
    })

    it("should load file data of new file after some file is loaded then removed and then the new file is loaded", async () => {
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload")
        const file = createXlsxFile(fileData, "example.xlsx")
        const newFile = createXlsxFile(newFileData, "new_example.xlsx")

        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );
        await user.click(getByLabelText("remove-file-to-load"))
        await user.upload(uploadField, newFile)
        await waitFor(() =>
            expect(getByText("new_example.xlsx")).toBeInTheDocument()
        );
        
        expect(container).toMatchSnapshot()
    })

    it("should load zipfile", async () => {           
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload-zip") 
        const zipContent = new Blob(["fake zip data"], { type: "application/zip" });
        const zipFile = new File([zipContent], "archive.zip", { type: "application/zip" });    

        await user.upload(uploadField, zipFile)
        await waitFor(() =>
            expect(getByText("archive.zip")).toBeInTheDocument()
        );

        expect(container).toMatchSnapshot()
    })

    it("should unload zipfile after remove archive file to upload button is clicked", async () => {           
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload-zip") 
        const zipContent = new Blob(["fake zip data"], { type: "application/zip" });
        const zipFile = new File([zipContent], "archive.zip", { type: "application/zip" });    

        await user.upload(uploadField, zipFile)
        await waitFor(() =>
            expect(getByText("archive.zip")).toBeInTheDocument()
        );

        await user.click(getByLabelText("remove-archive-file-to-load"))
        
        expect(container).toMatchSnapshot()
    })

    it("should reload archive file data after archive file is loaded then removed and loaded again", async () => {
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload-zip") 
        const zipContent = new Blob(["fake zip data"], { type: "application/zip" });
        const zipFile = new File([zipContent], "archive.zip", { type: "application/zip" });    

        await user.upload(uploadField, zipFile)
        await waitFor(() =>
            expect(getByText("archive.zip")).toBeInTheDocument()
        );
        await user.click(getByLabelText("remove-archive-file-to-load"))
        await user.upload(uploadField, zipFile)
        await waitFor(() =>
            expect(getByText("archive.zip")).toBeInTheDocument()
        );

        expect(container).toMatchSnapshot()
    })

    it("should load archive file data of new archive file after some archive file is loaded then removed and then the new archive file is loaded", async () => {
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload-zip") 
        const zipContent = new Blob(["fake zip data"], { type: "application/zip" });
        const zipFile = new File([zipContent], "archive.zip", { type: "application/zip" });
        const newZipContent = new Blob(["new fake zip data"], { type: "application/zip" });
        const newZipFile = new File([zipContent], "new_archive.zip", { type: "application/zip" });

        await user.upload(uploadField, zipFile)
        await waitFor(() =>
            expect(getByText("archive.zip")).toBeInTheDocument()
        );
        await user.click(getByLabelText("remove-archive-file-to-load"))
        await user.upload(uploadField, newZipFile)
        await waitFor(() =>
            expect(getByText("new_archive.zip")).toBeInTheDocument()
        );
        
        expect(container).toMatchSnapshot()
    })

    it("should have import collection button disabled if file was not loaded", async () => {           
        const {getByLabelText, queryByText} = renderComponent()
        const importCollectionButton = getByLabelText("import-data")
        const collectionNameInputField = getByLabelText("name")
        const descriptionInputField = getByLabelText("description")

        await user.type(collectionNameInputField, exampleCollectionData.name)
        await user.type(descriptionInputField, exampleCollectionData.description)
        await user.click(importCollectionButton)
        expect(importCollectionButton).toBeDisabled()

        expect(queryByText(/nazwa kolekcji jest wymagana/i)).not.toBeInTheDocument()
        expect(queryByText(/opis kolekcji jest wymagany/i)).not.toBeInTheDocument()
    })

    it("should show collection name is required error if collection name was not provided before clicking import collection button", async () => {           
        const {getByRole, getByText, getByLabelText, queryByText} = renderComponent()
        const file = createXlsxFile(fileData, "example.xlsx")      
        const uploadField = getByLabelText("upload")
        const descriptionInputField = getByLabelText("description")
        const importCollectionButton = getByLabelText("import-data")
        
        await user.upload(uploadField, file)
        await user.type(descriptionInputField, exampleCollectionData.description)
        await user.click(importCollectionButton)

        expect(getByText(/nazwa kolekcji jest wymagana/i)).toBeInTheDocument()
        expect(queryByText(/nie wczytano pliku/i)).not.toBeInTheDocument()
        expect(queryByText(/opis kolekcji jest wymagany/i)).not.toBeInTheDocument()
    })

    it("should show collection description is required error if description was not provided before clicking import collection button", async () => {           
        const {getByRole, getByText, queryByText, getByLabelText} = renderComponent()
        const file = createXlsxFile(fileData, "example.xlsx")      
        const uploadField = getByLabelText("upload")
        const nameInputField = getByLabelText("name")
        const importCollectionButton = getByLabelText("import-data")

        await user.upload(uploadField, file)
        await user.type(nameInputField, exampleCollectionData.name)
        await user.click(importCollectionButton)

        expect(getByText(/opis kolekcji jest wymagany/i)).toBeInTheDocument()
        expect(queryByText(/nie wczytano pliku/i)).not.toBeInTheDocument()
        expect(queryByText(/nazwa kolekcji jest wymagana/i)).not.toBeInTheDocument()
    })

    it("should call importDataAsCollection with correct arguments when import collection button is clicked", async () => {           
        const {getByText, getByLabelText} = renderComponent()
        const file = createXlsxFile(fileData, "example.xlsx")      
        const uploadField = getByLabelText("upload")
        const zipUploadField = getByLabelText("upload-zip") 
        const zipContent = new Blob(["fake zip data"], { type: "application/zip" });
        const zipFile = new File([zipContent], "archive.zip", { type: "application/zip" });    
        const nameInputField = getByLabelText("name")
        const descriptionInputField = getByLabelText("description")
        const importCollectionButton = getByLabelText("import-data")

        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );
        await user.type(nameInputField, exampleCollectionData.name)
        await user.type(descriptionInputField, exampleCollectionData.description)
        await user.upload(zipUploadField, zipFile)
        await waitFor(() =>
            expect(getByText("archive.zip")).toBeInTheDocument()
        );
        await user.click(importCollectionButton)

        expect(mockImportDataAsCollection).toHaveBeenCalledWith(
            fileData,
            exampleCollectionData.name,
            exampleCollectionData.description,
            jwtToken,
            zipFile
        )
    })

    it("should call importData with correct parameters when id and filename columns are included in spreadsheet file", async () => {           
        const {getByLabelText, getByText, container} = renderComponent()
        const uploadField = getByLabelText("upload")
        const file = createXlsxFile(fileDataWithIdsAndFilenames, "example.xlsx")
        const nameInputField = getByLabelText("name")
        const descriptionInputField = getByLabelText("description")

        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );
        await user.type(nameInputField, exampleCollectionData.name)
        await user.type(descriptionInputField, exampleCollectionData.description)
        
        expect(container).toMatchSnapshot()

        await user.click(getByLabelText("import-data"))

        expect(mockImportDataAsCollection).toHaveBeenCalledWith(
            fileDataWithIdsAndFilenames,
            exampleCollectionData.name,
            exampleCollectionData.description,
            jwtToken,
            undefined
        )
    })

    it("should change select option when another option is selected by user and call importDataAsCollection with correct arguments when import collection button is clicked", async () => {           
        const {getByLabelText, getByText} = renderComponent()
        const file = createXlsxFile(fileData, "example.xlsx")      
        const uploadField = getByLabelText("upload")
        const importCollectionButton = getByLabelText("import-data")
        const fileDataModified = JSON.parse(JSON.stringify(fileData))
        fileDataModified[0][3] = "Title.Subtitle.Subsubtitle.Artists"

        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );

        const artistsParentSelect = getByLabelText("Artists-parent-select")
        expect(artistsParentSelect).toHaveValue("-")
        await user.selectOptions(artistsParentSelect, "Subsubtitle")
        expect(artistsParentSelect).toHaveValue("Subsubtitle")
        await user.click(importCollectionButton)
        
        expect(mockImportDataAsCollection).toHaveBeenCalledWith(fileDataModified, "", "", jwtToken, undefined)
    })

    it("should show circular references error when options selected by users cause them", async () => {           
        const {getByText, getByLabelText} = renderComponent()
        const file = createXlsxFile(fileData, "example.xlsx")      
        const uploadField = getByLabelText("upload")
        const importCollectionButton = getByLabelText("import-data")
        const fileDataModified = JSON.parse(JSON.stringify(fileData))
        fileDataModified[0][3] = "Title.Subtitle.Subsubtitle.Artists"

        await user.upload(uploadField, file)
        await waitFor(() =>
            expect(getByText("example.xlsx")).toBeInTheDocument()
        );

        const subsubtitleParentSelect = getByLabelText("Title-parent-select")

        await user.selectOptions(subsubtitleParentSelect, "Subsubtitle")

        expect(subsubtitleParentSelect).toHaveValue("Subsubtitle")
        expect(getByText(/wykryto cykliczne referencje/i)).toBeInTheDocument()
        expect(importCollectionButton).toBeDisabled()
    })

    it.each([
        {axiosError: {response: {data: {error: "Incorrect request body provided"}}}},
        {axiosError: {response: {data: {error: "Invalid data in the spreadsheet file", cause: "Error: Header has duplicate values"}}}},
        {axiosError: {response: {data: {error: "Invalid data in the spreadsheet file", cause: "Error: Row contains more columns than the header"}}}},
        {axiosError: {response: {data: {error: "Invalid data in the spreadsheet file", cause: "Error: Header has empty fields"}}}},
        {axiosError: {response: {data: {error: "Invalid data in the spreadsheet file", cause: "Error: No subcategory name after the dot symbol in header field: Title"}}}},
        {axiosError: {response: {data: {error: "Invalid data in the spreadsheet file", cause: "Error: Missing parent category: Title"}}}},
        {axiosError: {response: {data: {error: "Invalid categories data", cause: `BrakujƒÖce kategorie nadrzƒôdne: Title`}}}},
        {axiosError: {response: {data: {error: "Database unavailable"}}}},
        
        
    ])(`should show appropriate error message when importData rejects with error cause - $axiosError.response.data.cause`,
        async ({axiosError}) => {
            mockImportDataAsCollection.mockImplementation(() => {
                throw axiosError;
            });
            const {getByLabelText, getByText} = renderComponent()
            const uploadField = getByLabelText("upload")
            const file = createXlsxFile(fileData, "example.xlsx")
            const nameInputField = getByLabelText("name")
            const descriptionInputField = getByLabelText("description")

            await user.upload(uploadField, file)
            await waitFor(() =>
                expect(getByText("example.xlsx")).toBeInTheDocument()
            );
            await user.type(nameInputField, exampleCollectionData.name)
            await user.type(descriptionInputField, exampleCollectionData.description)
            await user.click(getByLabelText("import-data"))
            
            expect(getByLabelText("server-error")).toMatchSnapshot()	
        }
    )
})