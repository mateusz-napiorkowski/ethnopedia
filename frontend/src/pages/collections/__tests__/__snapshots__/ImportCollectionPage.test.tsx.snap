// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportCollectionPage tests should render initial state 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-50 dark:bg-gray-900 flex flex-col"
  >
    <nav
      class="bg-gray-800 border-gray-200 dark:bg-gray-200 w-full border-b dark:border-gray-600"
    >
      <div
        class="max-w-screen-xl flex flex-wrap justify-between mx-auto p-1"
      >
        <div
          class="flex items-center mt-1"
        >
          <h1
            class="logo text-gray-200 dark:text-gray-900 text-6xl "
          >
            ethnopedia
          </h1>
        </div>
        <div
          class="flex items-center md:order-2"
        >
          <button
            class="text-gray-800 bg-white border border-gray-300 focus:outline-none
                           hover:bg-gray-100 focus:ring-4 focus:ring-gray-200 font-medium rounded-lg text-sm
                           p-2.5 mr-2 dark:bg-gray-800 dark:text-white dark:border-gray-600
                           dark:hover:bg-gray-700 dark:hover:border-gray-600 dark:focus:ring-gray-700"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="w-5 h-5 text-gray-800 dark:text-gray-200"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            class="text-gray-800 bg-white border border-gray-300 focus:outline-none
                           hover:bg-gray-100 focus:ring-4 focus:ring-gray-200 font-medium rounded-lg text-sm
                           p-2.5 mr-2 dark:bg-gray-800 dark:text-white dark:border-gray-600
                           dark:hover:bg-gray-700 dark:hover:border-gray-600 dark:focus:ring-gray-700"
            type="button"
          >
            <svg
              class="w-4 h-4 m-0.5"
            />
          </button>
          <div
            aria-label="show-dropdown"
            class="rounded-full bg-gray-300
                        p-3 cursor-pointer"
          >
            <svg />
          </div>
          <h2
            class="font-semibold ml-2 text-lg dark:text-gray-800 text-white"
          >
            123
          </h2>
        </div>
      </div>
    </nav>
    <div
      class="container mx-auto px-24 sm:px-32 md:px-40 lg:px-48 mt-4 max-w-screen-lg"
    >
      <nav
        class="flex"
      >
        <ol
          class="inline-flex items-center space-x-1 md:space-x-2"
        >
          <li
            class="inline-flex items-center text-lg"
          >
            <a
              class="inline-flex items-center font-medium text-gray-600 dark:text-gray-400 hover:text-blue-600 dark:hover:text-blue-600"
              href="/"
            >
              Strona główna
            </a>
          </li>
          <span
            class="self-center"
          >
            <svg
              class="w-3 h-3 text-gray-600 dark:text-gray-400"
            />
          </span>
          <li
            class="inline-flex items-center text-lg"
          >
            <a
              class="inline-flex items-center font-medium text-gray-600 dark:text-gray-400 hover:text-blue-600 dark:hover:text-blue-600"
              href="/import-collection"
            >
              Importuj kolekcję
            </a>
          </li>
        </ol>
      </nav>
      <div
        class="mt-4"
      >
        <div
          class="bg-white dark:bg-gray-800 rounded-lg shadow-md border dark:border-gray-600 p-8"
        >
          <div
            class="flex items-start rounded-t border-b pb-2"
          >
            <h3
              class="text-2xl font-bold text-gray-900 dark:text-white"
            >
              Importuj kolekcję
            </h3>
          </div>
          <form>
            <label
              class="block text-sm font-bold text-gray-700 dark:text-white my-2"
              for="dropzone-file"
            >
              Plik
            </label>
            <label
              aria-label="upload"
              class="flex flex-col items-start justify-start p-2 border-2 border-gray-200
                                    border-dashed rounded-lg cursor-pointer bg-gray-50 dark:hover:bg-gray-600
                                    dark:bg-gray-800 hover:bg-gray-100 dark:border-gray-600
                                    dark:hover:border-gray-500 dark:hover:bg-gray-700"
              for="dropzone-file"
            >
              <div
                class="flex flex-row items-center justify-center gap-4"
              >
                <svg
                  class="w-12 h-12 text-gray-500 dark:text-gray-400"
                />
                <p
                  class="text-sm font-normal text-gray-500 dark:text-gray-400"
                >
                  Kliknij, aby przesłać plik
                </p>
              </div>
              <input
                class="hidden"
                id="dropzone-file"
                type="file"
              />
            </label>
            <p
              class="block text-sm font-semibold text-gray-700 dark:text-white my-2"
            >
               
            </p>
            <label
              class="block text-sm font-bold text-gray-700 dark:text-white my-2"
              for="name"
            >
              Nazwa
            </label>
            <textarea
              aria-label="name"
              class="w-full px-4 py-2 border rounded-lg text-gray-700 focus:outline-none"
              id="name"
              name="name"
            />
            <div
              class="text-red-500 text-sm"
            >
               
            </div>
            <label
              class="block text-sm font-bold text-gray-700 dark:text-white my-2"
              for="description"
            >
              Opis
            </label>
            <textarea
              aria-label="description"
              class="w-full px-4 py-2 border rounded-lg text-gray-700 focus:outline-none"
              id="description"
              name="description"
            />
            <div
              class="text-red-500 text-sm mb-2"
            >
               
            </div>
            <hr />
            <div
              class="text-red-500 text-sm"
            >
               
            </div>
            <div
              class="text-red-500 text-sm h-15 overflow-y-scroll"
            />
            <div
              class="text-red-500 text-sm"
            >
               
            </div>
            <div
              class="flex justify-end mt-6"
            >
              <button
                class="px-4 py-2 mr-2"
                type="button"
              >
                Anuluj
              </button>
              <button
                class="px-4 py-2 color-button"
                disabled=""
                type="submit"
              >
                Importuj kolekcję
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ImportCollectionPage tests should update upload component and show loaded categories configuration menu with correct initial category structure after file is loaded 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-50 dark:bg-gray-900 flex flex-col"
  >
    <nav
      class="bg-gray-800 border-gray-200 dark:bg-gray-200 w-full border-b dark:border-gray-600"
    >
      <div
        class="max-w-screen-xl flex flex-wrap justify-between mx-auto p-1"
      >
        <div
          class="flex items-center mt-1"
        >
          <h1
            class="logo text-gray-200 dark:text-gray-900 text-6xl "
          >
            ethnopedia
          </h1>
        </div>
        <div
          class="flex items-center md:order-2"
        >
          <button
            class="text-gray-800 bg-white border border-gray-300 focus:outline-none
                           hover:bg-gray-100 focus:ring-4 focus:ring-gray-200 font-medium rounded-lg text-sm
                           p-2.5 mr-2 dark:bg-gray-800 dark:text-white dark:border-gray-600
                           dark:hover:bg-gray-700 dark:hover:border-gray-600 dark:focus:ring-gray-700"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="w-5 h-5 text-gray-800 dark:text-gray-200"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            class="text-gray-800 bg-white border border-gray-300 focus:outline-none
                           hover:bg-gray-100 focus:ring-4 focus:ring-gray-200 font-medium rounded-lg text-sm
                           p-2.5 mr-2 dark:bg-gray-800 dark:text-white dark:border-gray-600
                           dark:hover:bg-gray-700 dark:hover:border-gray-600 dark:focus:ring-gray-700"
            type="button"
          >
            <svg
              class="w-4 h-4 m-0.5"
            />
          </button>
          <div
            aria-label="show-dropdown"
            class="rounded-full bg-gray-300
                        p-3 cursor-pointer"
          >
            <svg />
          </div>
          <h2
            class="font-semibold ml-2 text-lg dark:text-gray-800 text-white"
          >
            123
          </h2>
        </div>
      </div>
    </nav>
    <div
      class="container mx-auto px-24 sm:px-32 md:px-40 lg:px-48 mt-4 max-w-screen-lg"
    >
      <nav
        class="flex"
      >
        <ol
          class="inline-flex items-center space-x-1 md:space-x-2"
        >
          <li
            class="inline-flex items-center text-lg"
          >
            <a
              class="inline-flex items-center font-medium text-gray-600 dark:text-gray-400 hover:text-blue-600 dark:hover:text-blue-600"
              href="/"
            >
              Strona główna
            </a>
          </li>
          <span
            class="self-center"
          >
            <svg
              class="w-3 h-3 text-gray-600 dark:text-gray-400"
            />
          </span>
          <li
            class="inline-flex items-center text-lg"
          >
            <a
              class="inline-flex items-center font-medium text-gray-600 dark:text-gray-400 hover:text-blue-600 dark:hover:text-blue-600"
              href="/import-collection"
            >
              Importuj kolekcję
            </a>
          </li>
        </ol>
      </nav>
      <div
        class="mt-4"
      >
        <div
          class="bg-white dark:bg-gray-800 rounded-lg shadow-md border dark:border-gray-600 p-8"
        >
          <div
            class="flex items-start rounded-t border-b pb-2"
          >
            <h3
              class="text-2xl font-bold text-gray-900 dark:text-white"
            >
              Importuj kolekcję
            </h3>
          </div>
          <form>
            <label
              class="block text-sm font-bold text-gray-700 dark:text-white my-2"
              for="dropzone-file"
            >
              Plik
            </label>
            <label
              aria-label="upload"
              class="flex flex-col items-start justify-start p-2 border-2 border-gray-200
                                    border-dashed rounded-lg cursor-pointer bg-gray-50 dark:hover:bg-gray-600
                                    dark:bg-gray-800 hover:bg-gray-100 dark:border-gray-600
                                    dark:hover:border-gray-500 dark:hover:bg-gray-700"
              for="dropzone-file"
            >
              <div
                class="flex flex-row items-center justify-center gap-4"
              >
                <svg
                  class="w-12 h-12"
                />
                <p
                  class="text-sm font-bold text-gray-500 dark:text-gray-400"
                >
                  example.xlsx
                </p>
              </div>
              <input
                class="hidden"
                id="dropzone-file"
                type="file"
              />
            </label>
            <p
              class="block text-sm font-semibold text-gray-700 dark:text-white my-2"
            >
              <span>
                Liczba rekordów: 
                <span
                  class="font-normal"
                >
                  1
                </span>
              </span>
            </p>
            <label
              class="block text-sm font-bold text-gray-700 dark:text-white my-2"
              for="name"
            >
              Nazwa
            </label>
            <textarea
              aria-label="name"
              class="w-full px-4 py-2 border rounded-lg text-gray-700 focus:outline-none"
              id="name"
              name="name"
            />
            <div
              class="text-red-500 text-sm"
            >
              Nazwa kolekcji jest wymagana
            </div>
            <label
              class="block text-sm font-bold text-gray-700 dark:text-white my-2"
              for="description"
            >
              Opis
            </label>
            <textarea
              aria-label="description"
              class="w-full px-4 py-2 border rounded-lg text-gray-700 focus:outline-none"
              id="description"
              name="description"
            />
            <div
              class="text-red-500 text-sm mb-2"
            >
              Opis kolekcji jest wymagany
            </div>
            <hr />
            <p
              class="text-sm font-bold text-gray-700 dark:text-white my-2"
            >
              Wczytane kategorie
            </p>
            <p
              class="text-sm font-normal text-gray-700 dark:text-white my-2"
            >
              Sprawdz, czy kategorie zostały poprawnie wczytane. Zaznacz, które kategorie są główne, a które są podkategoriami.
            </p>
            <div
              class="flex flex-col items-center justify-center mt-4 mb-4 p-4 border-2 border-gray-200
                                    border-solid rounded-lg bg-gray-50 dark:hover:bg-gray-600
                                    dark:bg-gray-800 dark:border-gray-600
                                    dark:hover:border-gray-500"
            >
              <div
                class="flex flex-row w-full items-start justify-start"
              >
                <span
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                >
                  Kategoria:
                </span>
                <span
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                >
                  Kategoria nadrzędna:
                </span>
              </div>
              <div
                class="flex flex-row w-full items-start justify-start"
              >
                <label
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  for="Title-parent"
                >
                  Title
                </label>
                <select
                  aria-label="Title-parent-select"
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  id="Title-parent"
                >
                  <option
                    value="-"
                  >
                    -
                  </option>
                  <option
                    value="Subtitle"
                  >
                    Subtitle
                  </option>
                  <option
                    value="Subsubtitle"
                  >
                    Subsubtitle
                  </option>
                  <option
                    value="Artists"
                  >
                    Artists
                  </option>
                </select>
              </div>
              <div
                class="flex flex-row w-full items-start justify-start"
              >
                <label
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  for="Subtitle-parent"
                >
                  Subtitle
                </label>
                <select
                  aria-label="Subtitle-parent-select"
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  id="Subtitle-parent"
                >
                  <option
                    value="-"
                  >
                    -
                  </option>
                  <option
                    value="Title"
                  >
                    Title
                  </option>
                  <option
                    value="Subsubtitle"
                  >
                    Subsubtitle
                  </option>
                  <option
                    value="Artists"
                  >
                    Artists
                  </option>
                </select>
              </div>
              <div
                class="flex flex-row w-full items-start justify-start"
              >
                <label
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  for="Subsubtitle-parent"
                >
                  Subsubtitle
                </label>
                <select
                  aria-label="Subsubtitle-parent-select"
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  id="Subsubtitle-parent"
                >
                  <option
                    value="-"
                  >
                    -
                  </option>
                  <option
                    value="Title"
                  >
                    Title
                  </option>
                  <option
                    value="Subtitle"
                  >
                    Subtitle
                  </option>
                  <option
                    value="Artists"
                  >
                    Artists
                  </option>
                </select>
              </div>
              <div
                class="flex flex-row w-full items-start justify-start"
              >
                <label
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  for="Artists-parent"
                >
                  Artists
                </label>
                <select
                  aria-label="Artists-parent-select"
                  class="block w-1/2 text-sm font-semibold text-gray-700 dark:text-white my-2"
                  id="Artists-parent"
                >
                  <option
                    value="-"
                  >
                    -
                  </option>
                  <option
                    value="Title"
                  >
                    Title
                  </option>
                  <option
                    value="Subtitle"
                  >
                    Subtitle
                  </option>
                  <option
                    value="Subsubtitle"
                  >
                    Subsubtitle
                  </option>
                </select>
              </div>
            </div>
            <div
              class="text-red-500 text-sm"
            >
               
            </div>
            <div
              class="text-red-500 text-sm h-15 overflow-y-scroll"
            />
            <div
              class="text-red-500 text-sm"
            >
               
            </div>
            <div
              class="flex justify-end mt-6"
            >
              <button
                class="px-4 py-2 mr-2"
                type="button"
              >
                Anuluj
              </button>
              <button
                class="px-4 py-2 color-button"
                type="submit"
              >
                Importuj kolekcję
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
